### PR TITLE
[maskedtensor] add basic tests and unary/binary/reduction tests from common_method_invocations

### PR DESCRIPTION
**Pull request by george-qi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84887
* __->__ #82841

Decided offline on the invariant that:

`masked_tensor` calls `MaskedTensor()`, which is analogous to `torch.tensor`
`as_masked_tensor` calls `MaskedTensor._from_values()`, which is analogous to `torch.as_tensor`

### CODE REVIEW

File: test/test_maskedtensor.py [hunk 1]
```diff
@@ -96,6 +105,195 @@ def _fix_fn_name(fn_name):
     return fn_name
 
 
+class TestBasics(TestCase):
+    def test_add(self):
```
Comment (bhosmer) [inline test/test_maskedtensor.py:109]: this is a negative test of (I think) incompatible masks, name needs to reflect what is being tested. also idk if this means add() isn't being positive tested but if not that should be added

File: test/test_maskedtensor.py [hunk 2]
```diff
@@ -96,6 +105,195 @@ def _fix_fn_name(fn_name):
...
+                [True, True, True, False],
+                [False, True, False, True],
+                [True, True, False, False],
+            ]
+        )
+        mx = MaskedTensor(x, m, requires_grad=True)
+        ts = torch.softmax(mx, -1)
+        ts.sum().backward()
+        xinf = x.masked_fill(~m, float("-inf")).detach().clone().requires_grad_()
+        torch.softmax(xinf, -1)
```
Comment (bhosmer) [inline test/test_maskedtensor.py:129]: what is being tested here

File: test/test_maskedtensor.py [hunk 3]
```diff
@@ -96,6 +105,195 @@ def _fix_fn_name(fn_name):
...
+
+    def test_where(self):
+        # http://pytorch.org/maskedtensor/main/notebooks/nan_grad.html
+        x = torch.tensor([-10.0, -5, 0, 5, 10, 50, 60, 70, 80, 90, 100])
+        mask = x < 0
+        mx = MaskedTensor(x, mask, requires_grad=True)
+        my = MaskedTensor(torch.ones_like(x), ~mask, requires_grad=True)
+        y = torch.where(mask, torch.exp(mx), my)
+        s = y.sum()
+        s.backward()
```
Comment (bhosmer) [inline test/test_maskedtensor.py:139]: what is being tested here

File: test/test_maskedtensor.py [hunk 4]
```diff
@@ -96,6 +105,195 @@ def _fix_fn_name(fn_name):
...
+
+    def test_mha_issue_41508(self):
```
Comment (bhosmer) [inline test/test_maskedtensor.py:141]: this isn't a sufficient test of the MT `multi_head_attention_forward()`. we need to either fully test it a la the core, non-MT version, or not ship it.

the fact that this is the only test makes me think that its purpose is basically a demo of how MTs can be used to fix 41508. That's cool but it doesn't motivate adding it to core - throwing yet another mha function in there is counterproductive, it just confuses things wrt what mha function to use. Personally I'd say we should remove it but @albanD @cpuhrsch might have other opinions.

File: test/test_maskedtensor.py [hunk 5]
```diff
@@ -96,6 +105,195 @@ def _fix_fn_name(fn_name):
...
+        # Tensors using zero and then cats them together.
+        # I don't know why the masks are coming into play here.
+        # It's an autograd thing.
+        k_data = torch.tensor([4.0])
+        k_mask = torch.tensor([True])
+        k = MaskedTensor(k_data[0], k_mask[0], requires_grad=True)
+        w = torch.tensor([1.0, 2.0], requires_grad=True)
+        w_q, w_k = w.chunk(2)
+        o0 = k + w_k
+        o0.backward()
```
Comment (bhosmer) [inline test/test_maskedtensor.py:192]: what is being tested

**Review by bhosmer — COMMENTED** [review]
quick first round of comments

File: test/test_maskedtensor.py (hunk 1 above)
Comment (george-qi) [reply test/test_maskedtensor.py:109]: added the positive test!

File: test/test_maskedtensor.py (hunk 2 above)
Comment (george-qi) [reply test/test_maskedtensor.py:129]: added an actual test for equivalence

File: test/test_maskedtensor.py (hunk 3 above)
Comment (george-qi) [reply test/test_maskedtensor.py:139]: added an equivalence test

File: test/test_maskedtensor.py (hunk 5 above)
Comment (george-qi) [reply test/test_maskedtensor.py]: removed

File: test/test_maskedtensor.py (hunk 4 above)
Comment (george-qi) [reply test/test_maskedtensor.py]: removed for now as per our offline discussion

File: test/test_maskedtensor.py [hunk 6]
```diff
@@ -281,84 +439,84 @@ def test_mean(self):
     """
         The following block of tests "test_mean_grad_case_1[a through e] are used to test the functionality of
         the two different ways of constructing MaskedTensors:
-            MaskedTensor(data, mask, requires_grad=True/False) -- NO differentiable constructor and always a leaf
-            MaskedTensor.from_values(data, mask) -- differentiable constructor
+            masked_tensor(data, mask, requires_grad=True/False) -- NO differentiable constructor and always a leaf
+            masked_tensor(data, mask) -- differentiable constructor
```
Comment (cpuhrsch) [inline test/test_maskedtensor.py:443]: Did you mean `as_masked_tensor` as in the piece of text below? Please check for this error within the PR and ask for review once you're confident it's been resolved.

**Review by cpuhrsch — CHANGES_REQUESTED** [review]

File: test/test_maskedtensor.py (hunk 6 above)
Comment (george-qi) [reply test/test_maskedtensor.py]: Yeah thanks, fixed

File: test/test_maskedtensor.py [hunk 7]
```diff
@@ -536,24 +692,154 @@ def test_masked_bmm(self):
             ]
         )
         attn_mask = attn_mask_bool.float().masked_fill_(attn_mask_bool, float("-inf"))
-        v = masked_bmm(x, x_mt.transpose(1, 2), attn_mask)
-        v.sum().backward()
 
-    def test_linear(self):
```
Comment (cpuhrsch) [inline test/test_maskedtensor.py:542]: What happened to this test?

File: torch/_masked/__init__.py [hunk 8]
```diff
@@ -1067,19 +1067,19 @@ def sum(
                 dtype = torch.int64
     dim_ = _canonical_dim(dim, input.ndim)
     mask_input = _combine_input_and_mask(sum, input, mask)
-    if input.layout == torch.strided:
+    if mask_input.layout == torch.strided:
```
Comment (cpuhrsch) [inline torch/_masked/__init__.py:1070]: Why change this?

File: torch/masked/maskedtensor/binary.py [hunk 9]
```diff
@@ -94,7 +94,10 @@ def _binary_helper(fn, args, kwargs, inplace):
     )
 
     args0_layout = data_args[0].layout
-    same_layout = args0_layout == data_args[1].layout
+    same_layout = (
+        (torch.is_tensor(data_args[1]) or is_masked_tensor(data_args[1])) and
```
Comment (cpuhrsch) [inline torch/masked/maskedtensor/binary.py:98]: Why is this additional check necessary?

File: torch/masked/maskedtensor/core.py [hunk 10]
```diff
@@ -342,7 +342,7 @@ def from_values(data, mask):
         class Constructor(torch.autograd.Function):
             @staticmethod
             def forward(ctx, data, mask):
-                return MaskedTensor(data.clone(), mask.clone())
+                return MaskedTensor(data, mask)
```
Comment (cpuhrsch) [inline torch/masked/maskedtensor/core.py:345]: Do we want to keep around `from_values`?

File: torch/masked/maskedtensor/reductions.py [hunk 11]
```diff
@@ -45,30 +46,35 @@ def _get_masked_fn(fn):
 def _torch_reduce_all(fn):
```
Comment (cpuhrsch) [inline torch/masked/maskedtensor/reductions.py:46]: As a general note on maskedtensor reductions, there already is a lot of code in torch/_masked. What are the plans for that code?

File: torch/masked/maskedtensor/binary.py (hunk 9 above)
Comment (george-qi) [reply torch/masked/maskedtensor/binary.py:98]: This is necessary because there are cases where we do `MaskedTensor + float`, and `float`s don't have a dtype so that would error out

File: torch/_masked/__init__.py (hunk 8 above)
Comment (george-qi) [reply torch/_masked/__init__.py:1070]: This was changed because the code was updated to support MaskedTensor inputs (in particular, if the input is MaskedTensor, then we want a differentiable `_combine_input_and_mask`, but we don't need one for vanilla tensor/masks. And so it's more accurate to check the `mask_input`'s layout, which is post-processing, rather than the input

File: torch/masked/maskedtensor/core.py (hunk 10 above)
Comment (george-qi) [reply torch/masked/maskedtensor/core.py:345]: I'm okay either way -- I suppose we could remove it, but I suppose also no harm. I suppose the more disciplined way would just have `masked_tensor` call `MaskedTensor()` and then `as_masked_tensor` call `MaskedTensor.from_values` -- I'll change it to this if we're all okay with that?

File: torch/masked/maskedtensor/reductions.py (hunk 11 above)
Comment (george-qi) [reply torch/masked/maskedtensor/reductions.py:46]: Discussed offline -- we'll move this into `torch/masked`

File: test/test_maskedtensor.py (hunk 7 above)
Comment (george-qi) [reply test/test_maskedtensor.py:542]: It wasn't actually testing an op on MaskedTensor afaik; it just performed `torch.nn.functional.linear` on a regular Tensor and then put it inside a `MaskedTensor`. Unless you had another test in mind to add?

File: test/test_maskedtensor.py [hunk 12]
```diff
@@ -96,6 +105,155 @@ def _fix_fn_name(fn_name):
...
+            ]
+        )
+        mt = masked_tensor(data, mask, requires_grad=True)
+        masked_res = torch.softmax(mt, -1)
+        masked_res.sum().backward()
+        xinf = data.masked_fill(~mask, float("-inf")).detach().clone().requires_grad_()
+        tensor_res = torch.softmax(xinf, -1)
+        tensor_res.sum().backward()
+        _compare_mt_t(masked_res, tensor_res)
+        if mt.grad is not None and xinf.grad is not None:
```
Comment (bhosmer) [inline test/test_maskedtensor.py:134]: not sure I get the reason for the conditional here

File: test/test_maskedtensor.py [hunk 13]
```diff
@@ -96,6 +105,155 @@ def _fix_fn_name(fn_name):
...
+        masked_res = torch.where(mask, torch.exp(mx), my)
+        masked_res.sum().backward()
+
+        x = data.detach().clone().requires_grad_()
+        y = torch.ones_like(x, requires_grad=True)
+        tensor_res = torch.where(mask, torch.exp(x), y)
+        tensor_res.sum().backward()
+
+        _compare_mt_t(mx.grad, x.grad)
+        _compare_mt_t(my.grad, y.grad)
```
Comment (bhosmer) [inline test/test_maskedtensor.py:152]: are we testing the primal here?

**Review by bhosmer — COMMENTED** [review]
quick comment pass, noted a couple basic things, will pick it up tomorrow

File: test/test_maskedtensor.py (hunk 12 above)
Comment (george-qi) [reply test/test_maskedtensor.py]: I think was an old check when the entire mask was masked out to be safe even though the entries should just be `nan`. Regardless, not necessary now for sure. Fixed

File: test/test_maskedtensor.py (hunk 13 above)
Comment (george-qi) [reply test/test_maskedtensor.py:153]: Makes sense to test as well, added

**Review by bhosmer — COMMENTED** [review]
Couple more basic issues but looks like we're getting close.

High level question: `TestBasics` are kind of spot checks, while `TestOperators` does the more thorough sweeps using `torch.testing` infra . But afaict only `TestBasics` does backward() testing. Is that typical?

File: test/test_maskedtensor.py [hunk 14]
```diff
@@ -536,24 +697,154 @@ def test_masked_bmm(self):
...
-                [False, False, False, True],
-                [False, True, True, True],
-                [False, True, False, True],
-            ]
-        )
-        x_mt = MaskedTensor(
-            x, ~(key_padding_mask.transpose(0, 1).unsqueeze(-1).expand_as(x))
-        )
+        mt_res = masked_bmm(x, x_mt.transpose(1, 2), attn_mask)
+        mt_res.sum().backward()  # make sure this runs
```
[the remark this answers is not in the document]
Comment (george-qi) [reply test/test_maskedtensor.py:702]: I've been thinking about this some more, and I might hav some fundamental misunderstanding here, but I don't believe the gradients should match.. please correct me if I misunderstood something cc @cpuhrsch since you originally authored `MaskedBmm` so likely know it better than I:

Consider the `MaskedBmm` implementation:

https://github.com/pytorch/pytorch/blob/3b9486461eac463771e45d6ecc0b3459cb199562/torch/masked/maskedtensor/matmul.py#L15-L41

and the derivative from `derivatives.yaml`:

https://github.com/pytorch/pytorch/blob/33352336b443dbfce1394f6b4950e8f33eff2cef/tools/autograd/derivatives.yaml#L348-L351

Indeed, `q_grad = torch.bmm(grad, k_transpose)` and `k_grad = torch.bmm(q_transpose, grad)`

In this testing, we've verified that `mt_res` is equivalent to `t_res` -- that is, the forward passes are equivalent, because we've masked out `k` when we call `x_masked = x.masked_fill(~x_mt.get_mask(), 0).detach().clone().requires_grad_()`

However, grad is not masked out in the backwards pass of vanilla `torch.bmm`, whereas it is implicitly in line 38 of the `MaskedBmm` implementation. 

In addition, instead of `torch.bmm(grad_data, k_trans)` (line 35 above) it should probably be `torch.bmm(grad.to_tensor(0), k_trans` -- is that right?

File: test/test_maskedtensor.py [hunk 15]
```diff
@@ -536,24 +697,154 @@ def test_masked_bmm(self):
...
+            mt_args = self._convert_mt_args(sample_args, mask, layout)
+
+            mt_result = op(mt, *mt_args, **sample_kwargs)
+            t_result = op(sample.input, *sample_args, **sample_kwargs)
+
+            _compare_mt_t(mt_result, t_result)
+
+            # If the operation is binary, check that lhs = masked, rhs = regular tensor also works
+            if is_binary(op):
+                mt_result2 = op(mt, *mt_args, **sample_kwargs)
```
[the remark this answers is not in the document]
Comment (george-qi) [reply test/test_maskedtensor.py]: Ugh sorry this was stale, thanks for the catch. Updated this; this only works when `torch.layout == torch.strided`, and of course `mt_args` should be `sample_args`.

**Comment by george-qi** [comment]
> Couple more basic issues but looks like we're getting close.
> 
> High level question: `TestBasics` are kind of spot checks, while `TestOperators` does the more thorough sweeps using `torch.testing` infra . But afaict only `TestBasics` does backward() testing. Is that typical?

Not entirely sure what's typical -- Christian and I have had more adhoc testing along these lines in the out-of-tree repoi, but I suppose part of migrating into core is adjusting to core's stricter testing, which is definitely a good thing :)

But as of this comment (because of your previous comment), I think every function in `TestBasics` had been changed to also testing the forward testing? e.g.:

https://github.com/pytorch/pytorch/blob/ea6c9c1131ae1282f95b89133504b9cef50d48bc/test/test_maskedtensor.py#L151

https://github.com/pytorch/pytorch/blob/ea6c9c1131ae1282f95b89133504b9cef50d48bc/test/test_maskedtensor.py#L220

and so on

File: test/test_maskedtensor.py (hunk 7 above)
Comment (cpuhrsch) [reply test/test_maskedtensor.py:542]: Does MaskedTensor support linear? If so we might need to add the test.

**Review by bhosmer — APPROVED** [review]

**Review by cpuhrsch — APPROVED** [review]
